Analysis/Tasks/PWGDQ/dileptonEE.cxx

```diff
@@ -0,0 +1,244 @@
+
+    DefineCuts();
+  }
+
+  void DefineCuts()
+  {
+    fEventCut = new AnalysisCompositeCut(true);
+
+    AnalysisCut* varCut = new AnalysisCut();
+    varCut->AddCut(VarManager::kVtxZ, -10.0, 10.0);
+    varCut->AddCut(VarManager::kVtxNcontrib, 0.5, 1e+10);
+
+    //TF1* cutLow = new TF1("cutLow", "pol1", 0., 0.1);
+    //cutLow->SetParameters(0.2635, 1.0);
+    //varCut->AddCut(VarManager::kVtxY, cutLow, 0.335, false, VarManager::kVtxX, 0.067, 0.070);
+
+    //varCut->AddCut(VarManager::kVtxY, 0.0, 0.335);
+    fEventCut->AddCut(varCut);
+
+    fTrackCut = new AnalysisCompositeCut(true); // true: use AND
+    AnalysisCut* cut1 = new AnalysisCut();
+    cut1->AddCut(VarManager::kPt, 0.15, 20.0);
+    cut1->AddCut(VarManager::kEta, -0.9, 0.9);
+    cut1->AddCut(VarManager::kTPCchi2, 0.0, 4.0);
+    cut1->AddCut(VarManager::kITSchi2, 0.0, 5.0);
+    cut1->AddCut(VarManager::kITSncls, 2.5, 7.5);
+    cut1->AddCut(VarManager::kTPCncls, 69.5, 159.5);
+    fTrackCut->AddCut(cut1);
+
+    VarManager::SetUseVars(AnalysisCut::fgUsedVars); // provide the list of required variables so that VarManager knows what to fill
+  }
+
+  void process(soa::Join<aod::ReducedEvents, aod::ReducedEventsExtended, aod::ReducedEventsVtxCov>::iterator event,
+               soa::Join<aod::ReducedTracks, aod::ReducedTracksBarrel, aod::ReducedTracksBarrelCov, aod::ReducedTracksBarrelPID> const& tracks)
+  {
+    // Reset the fgValues array
+    // TODO: reseting will have to be done selectively, for example run-wise variables don't need to be reset every event, but just updated if the run changes
+    //       The reset can be done selectively, using arguments in the ResetValues() function
+    VarManager::ResetValues();
+
+    VarManager::FillEvent<fgEventFillMap>(event);
+    fHistMan->FillHistClass("Event_BeforeCuts", VarManager::fgValues); // automatically fill all the histograms in the class Event
+    if (!fEventCut->IsSelected(VarManager::fgValues))
+      return;
+    fHistMan->FillHistClass("Event_AfterCuts", VarManager::fgValues);
+
+    // loop over barrel tracks and store positive and negative tracks in separate arrays
+    // TODO: use Partition initiaslized by vector of track indices when this will be available
+    std::vector<std::decay_t<decltype(tracks)>::iterator> selectedTracksPos;
+    std::vector<std::decay_t<decltype(tracks)>::iterator> selectedTracksNeg;
+    for (auto& track : tracks) {
+
+      VarManager::ResetValues(VarManager::kNEventWiseVariables, VarManager::kNMuonTrackVariables);
+
+      VarManager::FillTrack<fgTrackFillMap>(track);
+      fHistMan->FillHistClass("TrackBarrel_BeforeCuts", VarManager::fgValues);
+      if (!fTrackCut->IsSelected(VarManager::fgValues))
+        continue;
+      fHistMan->FillHistClass("TrackBarrel_AfterCuts", VarManager::fgValues);
+
+      if (track.charge() < 0)
```

Comment:
This can be simplified by using Partition<>, so that you do not have to accumulate a vector of iterators. You can declare it as
```
using tracktype = soa::Join<aod::ReducedTracks, aod::ReducedTracksBarrel, aod::ReducedTracksBarrelCov, aod::ReducedTracksBarrelPID>;
Partition<tracktype> posTracks = aod::reducedtrack::charge > 0;
```

Reply:
Hi @aalkin ,
Thank you for your comments.
Since I am still a beginner, please allow me to ask some questions.
First, histograms are filled with each track info.
Then, positive/negative tracks are separated for pairing after track selection.
So, I think track loop is still necessary here.

Reply:
Accumulating iterators instead of using tables directly most likely prevents optimizations. You can still use `isSelected` in the following double loop over positive and negative partitions.

Reply:
I see, thanks. Obviously, I have to learn more.
then,
using tracktype = soa::Join<aod::ReducedTracks, aod::ReducedTracksBarrel, aod::ReducedTracksBarrelCov, aod::ReducedTracksBarrelPID>;
should this be at the top of this macro?

Reply:
This is just a convenience alias, you can put it at the beginning of the file and use `tracktype` also in process() declaration.

Reply:
You are updating following these comments before we merge, right?

Reply:
I am trying to update this task.
But, if this is only for optimization and not crucial for hyperloop, please merge this now.
(many PRs seem stuck somehow, though.)
Then, I will continue to optimizing code.

Reply:
Hi Anton,
But is it possible to add tracks to a partition after the selection done in fTrackCut->IsSelected() ? If so, this would be great.
The separation over charges is the simple part. The main point is to have a selection cut which can be declared/defined outside the analysis class. This cut can contain criteria based on many derived variables (defined in the VarManager), which are not table columns. Both the number of variables and their limits can be defined by the user outside the analysis class in principle. How can i do that with Partition? I would like to avoid hardcoding in the task. As you maybe saw, in DQ we have histograms and cuts declared outside the analysis. If this is possible with Partition, would be great. We discussed once in a private channel on mattermost exactly about this and you said this is not possible yet, but maybe things changed in the meantime.
Cheers, Ionut

Reply:
Hi Ionut,

Currently - no, but looping over separated tracks and using the selection is still better than accumulating an array of iterators. It also may be better to have a multi-stage analysis - so that first task created a derived table with only selected tracks and subsequent tasks manipulate it in a declarative way. There is a PR in testing that will allow to use Configurables in filters and partitions. I understand, that currently the complete set of complicated cuts and selections cannot be implemented this way. However, any selections and partitions that _can_ would be very useful for performance.

The main point here is that iterators should not be treated as "track objects", otherwise there is no benefit of using these data structures at all.

Reply:
Hi Anton,
I think i understand what you mean. Note that actually what we do in DQ is to skim the data, so there is a task in front which runs before and produces a small dataset (the Reduced tables). Since these are electrons, we are looking here at about 1 per mille or less of the whole track sample when this task will run in production.
I agree on the performance issues, and we will try to improve our code. But would be also good to quantify these improvements because having user friendly code is also important and somewhere there should be an optimum.
In this particular case i am not sure it will be faster to apply the cut->IsSelected() in the 2-track loop (so applying this cut multiple times on the same track) than using the vector of iterators. But these optimizations can be definitely tried at some point.
In any case, i think that in order to make Partition to be adopted more easily, i think there should be more flexible ways of adding tracks to them. Not sure if possible, but maybe it can be discussed.
Cheers, Ionut

Reply:
In this particular case it may make sense to store the result of isSelected() in a column, join that column to your table, and then you can create a partition both on charge and the selection decision in one line.

Reply:
Hi Anton and Jan-Fiete,
Thank you for comments. I change my mind.
While I am updating this task, please keep this PR open (please do not merge yet).

@iarsene , Thank you for your help!

Reply:
Very good. I wait for a push to this PR then.

Reply:
Hello @jgrosseo and @aalkin ,
Thank you for comments. I implemented same way as https://github.com/AliceO2Group/AliceO2/pull/4291